Source/Diagnostics/Diagnostics.H

```diff
@@ -18,99 +18,84 @@ class Diagnostics
 {
 public:
     Diagnostics (int i, std::string name);
-    ~Diagnostics ();
+    /** Virtual Destructor to handle clean destruction of derived classes */
+    virtual ~Diagnostics () ;
     /** Pack (stack) all fields in the cell-centered output MultiFab m_mf_output.
-     * Fields are computed (e.g., cell-centered) on-the-fly using a functor.*/
+     * Fields are computed (e.g., cell-centered or back-transformed)
+       on-the-fly using a functor.*/
     void ComputeAndPack ();
     /** Flush m_mf_output and particles to file. */
-    void Flush ();
-    /** Flush raw data */
-    void FlushRaw ();
+    virtual void Flush (){};
+    /** Flush raw data (applies only to full diagnostics)*/
+    virtual void FlushRaw (){};
     /** Initialize pointers to main fields and allocate output multifab m_mf_output. */
-    void InitData ();
+    virtual void InitData (){};
     /** whether to flush at this time step
      * \param[in] step current time step
      * \param[in] force_flush if true, return true for any step
      * \return bool, whether to flush
      */
-    bool DoDump (int step, bool force_flush=false);
+    virtual bool DoDump (int step, bool force_flush=false){ return false;};
     /** Initialize functors that store pointers to the fields requested by the user.
-      * \param[in] lev level on which the vector of unique_ptrs to field functors is initialized.
-      */
-    void InitializeFieldFunctors (int lev);
+     * \param[in] lev level on which the vector of unique_ptrs to field functors is initialized.
+     */
+    virtual void InitializeFieldFunctors (int lev){};
```

Comment:
Ideally this function would be pure virtual because derived classes MUST implement it, so let's see if this is doable.
```suggestion
     *
     * Derived classes MUST implement this function, and it must allocate m_all_field_functors
     * and fill it with ComputeDiagFunctor objects.
     * \param[in] lev level on which the vector of unique_ptrs to field functors is initialized.
     */
    virtual void InitializeFieldFunctors (int lev) = 0;
```
An alternative is to implement it in Diagnostics directly, and have it call virtual functions (i.e., merge FullDiagnostics::InitializeFieldFunctors and BTDiagnostics::InitializeFieldFunctors).
The same applies for InitData

Reply:
the size of m_field_functors = m_varnames.size() for FullDiagnostics and =1 for BTDiagnostics.
Since it cannot be merged, I will make this a virtual function.